### DISCUSSION TITLE
Strong suspicion of performance regression due to re2

### DISCUSSION THREAD

**Discussion by onigoetz** [renovatebot/renovate#12722] [Request Help]
Hi,

I don't have definitive proof on the matter (probably tomorrow)

I have an install of Renovate that runs on more than 600 repositories, we upgrade monthly our instance and when doing the last update (29.3.3)
We saw that instead of taking 40 minutes to run all our 600 repositories it wasn't able to finish within one hour (did around 1/3 of all repositories). So around 4.5 times slower than the previous version. (27.21.0 in our case)

After a long process of narrowing down I was able to see that the Issue was present in version 28.11.1 but version 28.11.0 was running fine.

Which means that this was introduced here somewhere : https://github.com/renovatebot/renovate/releases/tag/28.11.1

My hunch is that with the introduction of `re2` and the wrapper function to transform RegExes a new RegEx is created on each function call : https://github.com/renovatebot/renovate/blob/main/lib/util/regex.ts#L20-L29 thus opting out of any JavaScript optimization.

I'll try tomorrow morning to disable re2 to see if the situation is improved or not.

What do you think of my findings ? should there be some kind of memoization on this function ?
I'd be happy to make a PR or test things on my install.

**Comment by rarkins** [comment]
Thank you for troubleshooting this! I would like to hear the results and support any attempt to improve it. Perhaps if we push these functions up to module scope we can revert most of the decrease.

**Reply by onigoetz** [reply]
Sure, I'll keep you updated as soon as I have some news

Yes, moving the RegEx to the module level was my first thought, but the amount of changes to make would be quite big, I thought that adding some memoization to the util could be a good first step. And then move the regEx to the module level.
Maybe there is also a linting rule that could help catch them to make sure there is no regression after that.

**Reply by onigoetz** [reply]
Okay so I made my tests and indeed that's the source of our slowness, I opened an issue here : https://github.com/renovatebot/renovate/issues/12736

**Comment by HonkingGoose** [comment]
I think this deserves to be an issue instead of a discussion. The benefit of having it as an issue is that is more visible for others, and we can track the progress better.

@rarkins Want me to convert this to an issue?

**Reply by onigoetz** [reply]
I opened it as a discussion first to see if that was something that was in your radar already and since my suspicions are not 100% confirmed yet, I didn't want to pollute issues with something that could (although unlikely at this stage) not be an issue in the end.

**Reply by HonkingGoose** [reply]
Oh okay, that's a reasonable thought! 😄 We can probably wait until you've found more proof to confirm your hunch before converting to an issue. 😉
